src/foam/nanos/auth/PropertyPermissionDAO.js

```diff
@@ -0,0 +1,133 @@
+foam.CLASS({
+    package: 'foam.nanos.auth',
+    name: 'PropertyPermissionDAO',
+  
+    javaImports: [
+      'foam.nanos.auth.AuthService',
+      'foam.core.FObject',
+      'foam.core.PropertyInfo',
+      'foam.core.X',
+      'foam.dao.ProxySink',
+      'java.util.Iterator',
+      'java.util.List'
+    ],
+  
+    extends: 'foam.dao.ProxyDAO',
+    documentation: `A DAO decorator that prevents user from updating / reading
+        properties for which the user does not have the update / read permission`,
+
+    methods: [
+      {
+        name: 'put_',
+        javaCode: `
+    FObject oldObj = getDelegate().find(obj.getProperty("id"));
+    return super.put_(x, hideProperties(x, obj, oldObj, true));
+        `,
+      },
+  
+      {
+        name: 'find_',
+        javaCode: `
+    FObject oldObj = getDelegate().find(id);
+    
+    if ( oldObj != null ) {
+      FObject copiedObj = oldObj.fclone();
+      return hideProperties(x, copiedObj, oldObj, false);
+    }
+
+    return super.find_(x, id);
+        `,
+      },
+
+      {
+        name: 'select_',
+        javaCode: `
+    if ( sink != null ) {
+      ProxySink proxySink = new ProxySink(x, sink) {
+        @Override
+        public void put(Object obj, foam.core.Detachable sub) {
+          FObject copiedObj = ((FObject) obj).fclone();
+          FObject oldObj = PropertyPermissionDAO.this.getDelegate().find(copiedObj.getProperty("id"));
+          if ( oldObj != null ) {
+            super.put(hideProperties(x, copiedObj, oldObj, false), sub);
+          } else {
+            super.put(obj, sub);
+          }
+        }
+      };
+
+      return ((ProxySink) super.select_(x, proxySink, skip, limit, order, predicate)).getDelegate();
+    }
+
+    return super.select_(x, sink, skip, limit, order, predicate);
+        `,
+      },
+
+      {
+        name: 'hideProperties',
```

Comment:
I don't think you should use the same method for filtering both put()'s and find/select()s as they do different things.

Reply:
Isn't most of the code the same for both put and find/select? So would it be better to use one method for put and a different one for find/select?